Correctly handle size and effective value of inputs.

turbolay opened the pull request:
This PR basically fixes the size issue (which was creating the inconsistency in fee rate) and uses the effective value of inputs in the Mixer (which is Wasabi's behavior). It also filters out inputs that costs less to create than their own value (again Wasabi's behavior)

Sorry for the diff

nopara73 commented on Sake/Mixer.cs:
```diff
@@ -32,7 +32,7 @@ public Mixer(FeeRate feeRate, Money minAllowedOutputAmount, Money maxAllowedOutp
             // Create many standard denominations.
             Denominations = CreateDenominations();
-            ChangeScriptType = GetNextScriptType();
+            ChangeScriptType = NBitcoinExtensions.GetNextScriptType(IsTaprootAllowed, Random);
```
The original code is in WalletWasabi's master. Unless you're planning to change that, this is not good.

nopara73 commented on Sake/Mixer.cs:
```diff
@@ -57,7 +57,7 @@ private IOrderedEnumerable<Output> CreateDenominations()
             Output CreateDenom(double sats)
             {
-                var scriptType = GetNextScriptType();
+                var scriptType = NBitcoinExtensions.GetNextScriptType(IsTaprootAllowed, Random);
```
The original code is in WalletWasabi's master. Unless you're planning to change that, this is not good.

nopara73 commented on Sake/Mixer.cs:
```diff
@@ -198,27 +198,27 @@ public IEnumerable<IEnumerable<ulong>> CompleteMix(IEnumerable<IEnumerable<ulong
         /// <param name="myInputsParam">Input effective values. The fee substracted, this is how the code works in the original repo.</param>
         /// <param name="maxVsizeCredentialValue">Maximum usable Vsize that client can get per alice.</param>
-        public IEnumerable<Output> Decompose(IEnumerable<Money> myInputsParam, IEnumerable<Output> denoms, int maxVsizeCredentialValue)
+        public IEnumerable<Output> Decompose(IEnumerable<Input> myInputsParam, IEnumerable<Output> denoms, int maxVsizeCredentialValue)
```
The original code in WalletWasabi's master uses no Input class. Unless you're planning to change that, this is not good.

nopara73 reviewed (changes requested):
I started reviewing it, but I'm stopping here, because I already see where this is going. It has introduced a lot of new divergence from WalletWasabi master. Can you redo this without introducing any divergence?

adamPetho commented on Sake/Program.cs:
```diff
@@ -6,39 +6,58 @@
 var feeRate = new FeeRate(10m);
 var random = new Random();
 
+var maxInputCost = Money.Satoshis(NBitcoinExtensions.P2trInputVirtualSize * feeRate.SatoshiPerByte);
```
How come that we calculate with `P2trInputVirtualSize`? 
AFAIK Segwit v0 Input costs more (Segwit v0 input 68 - Taproot input 57,5 according to [Rafe's calculation](https://github.com/zkSNACKs/WalletWasabi/issues/10675))
,so if we want to calculate the worst case scenario, we maybe want to calculate with Segwit v0 inputs? IDK.

turbolay replied:
Thank you! It was a mistake due to the previous state of the code.

turbolay commented:
@nopara73 I moved things around so you can better see that the only changes in `Decompose` are:
- Use `EffectiveValue` for inputs
- Correctly compute the `availableVSize`

But then I really don't know what you want me to do. WalletWasabi's master indeed doesn't use an `Input` class: it uses a `Credential` class to compute the `AvailableVSize` per `AliceClient` on the coordinator side.
The code on Sake's `master` regarding this is clearly outdated as it always considers an input as `P2WPKH`. It's also wrong as it ignores the fee in the amount of the inputs.

Sake's `master` only considers the `Amount` of the input, but in fact 3 values must be stored:
- The `ScriptType` of the inputs to match Wasabi's `master` and know their size and hence the fee they have to pay.
- Their `EffectiveValue` to use in `Decompose` (as in Wasabi's master and in all the comments in Sake)
- Their `Amount` to compute later on the `AnonSet` on the inputs

The `Input` class I introduced is a placeholder for these 3 values, if you don't want me to introduce it please suggest me an alternative.

nopara73 commented:
From the first glance this now looks good. If the `Decompose` method does not diverge more than it already does and it does not seem to do that anymore, then it looks good. This way we need not to touch WalletWasabi and so I can merge it more safely.

Anyhow, I'll review this in detail later.

nopara73 commented:
Both the output and the code looks good. Great work!
